internal/v2/cert_verifier/cert_verifier_test.go

```diff
@@ -35,39 +80,74 @@ func TestVerifyClientCertChain(t *testing.T) {
 	for _, tc := range []struct {
 		description              string
 		pool                     *x509.CertPool
-		expectedCommonName       string
 		rawCerts                 [][]byte
 		verifiedChains           [][]*x509.Certificate
-		err                      error
+		err			 error
 	}{
 		{
 			description: "static : empty certPool",
 			pool: certPool0,
-			expectedCommonName: "s2a_test_cert",
 			rawCerts: nil,
 			verifiedChains: nil,
-			err: nil,
+			err: errors.New("rpc error: code = Unknown desc = Client Peer Verification failed: client cert chain is empty."),
 		},
 		{
 			description: "static : non-empty certPool",
 			pool: certPool1,
-			expectedCommonName: "s2a_test_cert",
-			rawCerts: cert1.Certificate,
+			rawCerts: [][]byte{clientDERCert,},
 			verifiedChains: [][]*x509.Certificate{{cert1.Leaf},{}},
 			err: nil,
 		},
 	}{
 		t.Run(tc.description, func(t *testing.T) {
-			VerifyPeerCertificateFunc := VerifyClientCertificateChain(tc.expectedCommonName, tc.pool)
-			if got, want := VerifyPeerCertificateFunc(tc.rawCerts, tc.verifiedChains), tc.err; got != want {
-				t.Errorf("Peer Certificate verification failed: %v", got)
+			log.Printf("----TEST CASE: Verify Client Cert Chain: %s----", tc.description)
+			// Create new stream to S2Av2.
+			opts := []grpc.DialOption {
+				grpc.WithTransportCredentials(insecure.NewCredentials()),
+				grpc.WithReturnConnectionError(),
+				grpc.WithBlock(),
+			}
+			conn, err := grpc.Dial(address, opts...)
+			if err != nil {
+				log.Fatalf("Client: failed to connect: %v", err)
+			}
+			defer conn.Close()
+			c := s2av2pb.NewS2AServiceClient(conn)
+			log.Printf("Client: connected to: %s", address)
+			ctx, cancel := context.WithTimeout(context.Background(), defaultTimeout)
+			defer cancel()
+
+			// Setup bidrectional streaming session.
+			callOpts := []grpc.CallOption{}
+			cstream, err := c.SetUpSession(ctx, callOpts...)
+			if err != nil  {
+				log.Fatalf("Client: failed to setup bidirectional streaming RPC session: %v", err)
+			}
+			log.Printf("Client: set up bidirectional streaming RPC session.")
+
+			VerifyPeerCertificateFunc := VerifyClientCertificateChain(tc.pool, cstream)
+			got, want := VerifyPeerCertificateFunc(tc.rawCerts, tc.verifiedChains), tc.err
+			if ((got == nil) || (want == nil)) && (got != want) {
```

Comment:
I found this logic a little confusing - could we replace it with the following:

if want == nil {
   if got != nil {
      // error
   }
} else {
  if got == nil {
    // error
   }
   // compare error strings from got and want
}

wdyt?

Reply:
Similarly for the server side test.